### ISSUE TITLE
Slow with lot of dictionary items

### ISSUE THREAD

**Issue by enkelmedia** [marcemarc/tooorangey.DictionaryItemKeySearch#4]
Hi!

I have around 50 items now and are experiencing some performance issues, searching take some time. Is this expected?

**Comment by marcemarc** [comment]
Hi Markus, yes the search is using the Dictionary and Language classes from umbraco.cms.busineslogic, and worse that that each search it loops through the dictionary items, adds them to a list in memory, and then searches in memory (no caching!!) 

https://github.com/marcemarc/tooorangey.DictionaryItemKeySearch/blob/master/tooorangey.DictionaryItemKeySearch/tooorangey.DictionaryItemSearch.Api/Controllers/DictionarySearchController.cs

so I think we could revamp that a little bit and find some performance gains!!!

**Comment by enkelmedia** [comment]
It might be a good idea to store that collection that is created in a cache somewhere and kill that when a dictionary is changed? On the other side killing it when something changes would probably make each new search a "fresh hit" against the API as a search almost all the time would end up with a change of a dictionary item.

I'm not sure how they are stored in the db and "I know its not right" =D But it might be good for performance with a "LIKE"-query to the db.

**Comment by marcemarc** [comment]
Yes well now there is the new LocalizationService: https://github.com/umbraco/Umbraco-CMS/blob/5397f2c53acbdeb0805e1fe39fda938f571d295a/src/Umbraco.Core/Services/LocalizationService.cs

which looks like it will be a bit quicker... and benefit from some petapoco caching.

but ultimately there is no search, so possibly a pull request to core to add a Search to the DictionaryRepository:

https://github.com/umbraco/Umbraco-CMS/blob/5397f2c53acbdeb0805e1fe39fda938f571d295a/src/Umbraco.Core/Persistence/Repositories/DictionaryRepository.cs

maybe the neatest way to go.

Otherwise yes, caching, and an event hooking into dictionary item save to break the cache, would be easy and better than what is there.

**Comment by enkelmedia** [comment]
Hmm... whats the chance of getting the PR into the core? Could take years =D That would also be hard to support older versions of Umbraco. 

I'm willing to try to look at this but I would like to know which way you would perfer?

1. SQL-query to the DB (we could add a "fallback" that kicks in if the tables are changed in the future).
2. Cache items and search the cache

I think my vote is for 1 =D

**Comment by marcemarc** [comment]
Indeed, I think for the searching part, then the database is the most pragmatic approach:

  Select cmsLanguageText.pk, cmsLanguageText.value, cmsDictionary.pk, cmsDictionary.id, cmsDictionary.[key]
  from cmsLanguageText 
  Inner Join 
  cmsDictionary on cmsLanguageText.UniqueId = cmsDictionary.id
  where cmsLanguageText.value like '%else%' or cmsDictionary.[key] = 'ElsewhereOnTheWeb'

just need to ensure the syntax works with SQLCE

the updating of values via the dashboard should still be done with the services etc

cheers

Marc

**Comment by enkelmedia** [comment]
Cool! I'll spend some time on this during the next week(s) and get back when I have something working.
